h/static/scripts/directive/annotation.js

```diff
@@ -0,0 +1,701 @@
+    });
+    tags.store(newTags);
+
+    switch (vm.action) {
+      case 'create':
+        updateDomainModel(model, vm.annotation);
+        var onFulfilled = function() {
+          $rootScope.$emit('annotationCreated', model);
+          vm.view();
+        };
+        var onRejected = function(reason) {
+          flash.error(
+            errorMessage(reason), 'Saving annotation failed');
+        };
+        return model.$create().then(onFulfilled, onRejected);
+
+      case 'edit':
+        var updatedModel = angular.copy(model);
+        updateDomainModel(updatedModel, vm.annotation);
+        onFulfilled = function() {
+          angular.copy(updatedModel, model);
+          $rootScope.$emit('annotationUpdated', model);
+          vm.view();
+        };
+        onRejected = function(reason) {
+          flash.error(
+            errorMessage(reason), 'Saving annotation failed');
+        };
+        return updatedModel.$update({
+          id: updatedModel.id
+        }).then(onFulfilled, onRejected);
+    }
+  };
+
+  /**
+    * @ngdoc method
+    * @name annotation.AnnotationController#reply
+    * @description
+    * Creates a new message in reply to this annotation.
+    */
+  vm.reply = function() {
+    var id = model.id;
+    var references = model.references || [];
+
+    // TODO: Remove this check once we have server-side code to ensure that
+    // references is always an array of strings.
+    if (typeof references === 'string') {
+      references = [references];
+    }
+
+    references = references.concat(id);
+
+    var reply = annotationMapper.createAnnotation({
+      references: references,
+      uri: model.uri
+    });
+    reply.group = model.group;
+
+    if (session.state.userid) {
+      if (permissions.isShared(model.permissions, model.group)) {
+        reply.permissions = permissions.shared(reply.group);
+      } else {
+        reply.permissions = permissions.private();
+      }
+    }
+  };
+
+  /**
+    * @ngdoc method
+    * @name annotation.AnnotationController#render
+    * @description Called to update the view when the model changes.
+    */
+  vm.render = function() {
+    var draft = drafts.get(model);
+
+    // Extend the view model with a copy of the domain model.
+    // Note that copy is used so that deep properties aren't shared.
+    vm.annotation = angular.extend({}, angular.copy(model));
+
+    // If we have unsaved changes to this annotation, apply them
+    // to the view model.
+    if (draft) {
+      angular.extend(vm.annotation, angular.copy(draft));
+    }
+
+    vm.annotationURI = new URL(
+      '/a/' + vm.annotation.id, vm.baseURI).href;
+
+    // Extract the document metadata.
+    var uri = model.uri;
+    var domain = new URL(uri).hostname;
+    if (model.document) {
+      if (uri.indexOf('urn') === 0) {
+        var i;
+        for (i = 0; i < model.document.link.length; i++) {
+          var link = model.document.link[i];
+          if (!(link.href.indexOf('urn'))) {
```

Comment:
Strictly speaking, that could still be incorrect since the protocol could start with 'urn' but have something else after it, we should actually be checking for 'urn:' instead.